xcm/src/v1/multilocation.rs

```diff
@@ -329,36 +329,26 @@ impl MultiLocation {
 	///
 	/// Does not modify `self` in case of overflow.
 	pub fn reanchor(&mut self, target: &MultiLocation, ancestry: &MultiLocation) -> Result<(), ()> {
-		// TODO: https://github.com/paritytech/polkadot/issues/4489 Optimize this.
+		let mut target = target.clone();
+		let mut id = self.clone();
 
-		// 1. Use our `ancestry` to figure out how the `target` would address us.
-		let inverted_target = ancestry.inverted(target)?;
+		id.prepend_with(ancestry.clone()).map_err(|_| ())?;
+		target.prepend_with(ancestry.clone()).map_err(|_| ())?;
 
-		// 2. Prepend `inverted_target` to `self` to get self's location from the perspective of
-		// `target`.
-		self.prepend_with(inverted_target).map_err(|_| ())?;
+		// Remove common parents between now normalized id and target
+		while target.interior().first() == id.interior().first() {
+			target.take_first_interior();
+			id.take_first_interior();
+		}
 
-		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
-		// strictly needed.
-		self.simplify(target.interior());
+		id.parents = id
+			.parent_count()
+			.saturating_add(target.parent_count().saturating_add(target.interior().len() as u8));
```

Comment:
```suggestion
		let final_parents = (id.parent_count() as usize).saturating_add(
			(target.parent_count() as usize).saturating_add(target.interior().len()),
		);
		if final_parents > 255 {
			return Err(()) 
		}
		id.parents = final_parents as u8;
```

Reply:
Oh, I see. The idea is to return an error if there's an overflow and not just saturate it to fail silently? In that case I've switched to using a `checked_add` instead of `saturating_add`. That should give us the behaviour we want.